### ISSUE TITLE
Snapping points to the lines should be switchable

### ISSUE THREAD

**Issue by deevroman** [Zverik/every_door#541]
Reasons:
1. Sticking is sometimes unnecessary. For example, if the entryway is inside the archway of a house, it will stick to the outline of the house. Also relate to this #471
2. As long as sticking points to multi-polygons does not work #210, you may have the following situation: two adjacent buildings, one multi-polygon, the other polygon. Every Door in this case could snap a point to the building drawn by the line.
3. it is not clear which points snap to which ones. For example: if there is a footpath and a road next to each other, which one the barrier will be magnetized to.
4. Sniping is not obvious. As a new user, you can only find this out if you examine your changes after downloading. (or someone will tell you about it)

**Comment by gdabski** [comment]
See #541 for a more general issue.

**Comment by gdabski** [comment]
It would indeed be great to do something about it. Another example I have struggled with is snapping gates to highways. There were cases I intended to mark a gate on an already-mapped fence or wall. Instead the gate got snapped to the road that the fence was running along. We've even had navigation over a primary road broken for this reason after an unaware editor made an edit with EveryDoor :(

**Comment by mnalis** [comment]
Would it perhaps be possible so EveryDoor, **at the time the element is being added to the map**:

- asks the user _"Do you want to snap `node type` to `way type`"_ (e.g. _"Do you want to snap barrier=gate to highway=residential"_) and acts accordingly? 

- Or, alternatively, somehow visually indicate where the element is going to be snapped (like e.g. in StreetComplete _Street Parking overlay_, when it displays separate pin at exact location on a way where it is going to snap on, and does not show the pin if no snapping is going to take place because it is too far from the road)?

Because doing it invisibly to the user in the background at the time of upload is problematic as it breaks not only _"do not surprise the user"_ but also _"user should be in control what is being added to the map"_ maxims, and thus sometimes leads to unintended bad data.

**Comment by deevroman** [comment]
I don't think we need an additional step in the form of a dialog. A switch (or icon for entrance mode) in the object card would be enough.

**Comment by mnalis** [comment]
> I don't think we need an additional step in the form of a dialog

I agree that it sounds somewhat clunky if it pops up too often, but might be easiest to implement. I would much prefer the **direct visual indication mode** instead, which I mentioned earlier:

- like in [StreetComplete](https://wiki.openstreetmap.org/wiki/StreetComplete) _"parking overlay"_:
![small_Screenshot_20240522_121557_StreetComplete](https://github.com/Zverik/every_door/assets/156656/a02d2745-f8fb-4aa1-b404-f2a2806c21e3) ![small_Screenshot_20240522_121545_StreetComplete](https://github.com/Zverik/every_door/assets/156656/82028cc9-c2e8-475a-93fb-eaca62c8b38b)


- or like [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) _"insert node into way"_ long-press functionality:
   ![small_Screenshot_20240522_121637_SCEE](https://github.com/Zverik/every_door/assets/156656/a9c99500-453f-4e9c-a011-0e5b6eb4553f) ![small_Screenshot_20240522_121627_SCEE](https://github.com/Zverik/every_door/assets/156656/50dfa42f-abf4-4624-b04c-d732ffe8f091)

I however suspect that it is (maybe significantly) harder to implement, so that is why I suggested that only is second place (even if I would much prefer such solution, as such approach has already shown to work quite nicely and is easy to use) 

> A switch (or icon for entrance mode) in the object card would be enough.

I'm afraid manual snap on/off switch (or button like "door" icon in entrance mode) IMHO would **not** be good enough though, if I understood correctly what you meant there @deevroman ?

See [example above](https://github.com/Zverik/every_door/issues/541#issuecomment-1741820343)  by @gdabski or your point `3.` in original report. It would still leave it up to the heuristics (instead of the user) whether the gate would end on barrier or on the highway, which is a serious routing-breaking issue. And if you disabled snapping completely via some switch, then adding the gate would be quite useless.

Having a separate switch for each type of a way (e.g. "snap on a footway", "snap on a road", "snap on the barrier", ...) would not work either, not only because it sounds like an overkill (and usability issue), but also because it would not work on e.g. two `highway=residential` crossing and similar situations.

---

TL;DR: Visual indication **when and on what exactly the snapping will occur** would work best IMHO (as evidenced in StreetComplete/SCEE), and allow the user to opt-out of snapping by moving an element a little away. 

While adding additional manual switch to fully disable snapping _might_ be an useful ***addition*** (as it would allow for little more precise positioning without snapping), it is ***not suitable*** as a ***replacement*** for such visual indication (or some other solution putting user in charge of what exactly is being editing).

**Comment by Zverik** [comment]
Note to self: re-enable snapping plaques after I do this.

**Comment by juleskers** [comment]
New Mapper chiming in. I started mapping a couple of weeks ago, and appreciate EveryDoor for how easy it makes it to map hydrants and playgrounds (my current topic-of-interest).

I just learned about snapping, because I got a comment on my [playground-change-set](https://www.openstreetmap.org/changeset/181110023#map=19/49.929286/8.599878): "can the busses pass these fences?" 
I *intended* to place the playground-entrances, by the roadside, but because the fence isn't mapped, it seems the gates snapped onto the road/busroute?
